clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -1267,6 +1267,17 @@ CodeGenFunction::canTightenCheriBounds(llvm::Value *Value, QualType Ty,
           "referenced value is a weak symbol and could therefore be NULL");
     }
   }
+
+  // Don't set bounds on null pointer arithmetic used for hand-rolled offsetof
+  // idioms, even if it's UB.
+  const Expr *BaseDeref =
+    E->getDereferencedBaseExpr(getContext())->IgnoreParenCasts();
+  if (BaseDeref->isNullPointerConstant(getContext(),
+                                       Expr::NPC_ValueDependentIsNotNull))
+    return cannotSetBounds(*this, E, Ty, Kind,
```

Comment:
I wonder if we should be emit a warning and suggest replacing it with `(__builtin_)offsetof`?

Reply:
The thought did occur to me but then forgot about it; probably should